File: src/de/rub/bi/inf/openbimrl/functions/ifc/FilterByElement.java
package de.rub.bi.inf.openbimrl.functions.ifc;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.engine.ifc.IIFCModel;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;

import java.util.ArrayList;
import java.util.Collection;

/**
 * Filters a IFC model and retrieves all elements of a certain type.
 *
 * @author Marcel Stepien
 */
public class FilterByElement extends AbstractFunction {

    public FilterByElement(NodeProxy nodeProxy) {
        super(nodeProxy);
    }

    @Override
    public void execute(IIFCModel ifcModel) {

        String className = getInput(0);
        Class<?> theClass;
        try {
            // TODO fix
            theClass = Class.forName("com.apstex.ifctoolbox.ifc." + className);
            Collection<?> objects = ifcModel.getCollection(theClass);

            setResult(0, new ArrayList(objects));
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }

    }

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/FilterByElement.kt
package de.rub.bi.inf.openbimrl.functions.ifc

import de.rub.bi.inf.nativelib.IfcPointer
import de.rub.bi.inf.nativelib.NativeEngine
import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.NativeFunction
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionPort
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction

@OpenBIMRLFunction(
    description = "Filters a IFC model and retrieves all elements of a certain type.",
    inputs = [
        FunctionPort(0, "IfcType", String::class),
    ],
    outputs = [
        FunctionPort(0, "IfcElement_List", IfcPointer::class, isCollection = true),
    ],
)
class FilterByElement(nodeProxy: NodeProxy) : NativeFunction(nodeProxy) {
    override fun executeNative() {
        val ifcType = getInput<String>(0) ?: return
        setResult(0, IfcPointer.fromHandles(NativeEngine.filterByElement(ifcType)))
    }
}
